File: src/typescript/load-test.ts
import * as BBPromise from "bluebird";
// import { v1 as newUuid } from "node-uuid";
import { DeviceClient } from "./DeviceClient";
import * as stopwatch from "./Stopwatch";

class Device {
    private static nextID = 0;
    private readonly id = `device-${Device.nextID++}`; // newUuid();
    private readonly client: DeviceClient;

    private static readonly MinTemp = -10;
    private static readonly MaxTemp = 45;

    constructor(serverAddress: string) {
        this.client = new DeviceClient(serverAddress);
    }

    async run(attempts: number, secondsBetweenAttemps: number) {
        const tempSvc = this.client.temperatureService;
        let celcius = this.randomTemperature();

        await this.setTemperature(celcius);

        const tempStream = tempSvc.getTemperature({device: this.id});
        const dataStopwatch = stopwatch.startStopwatch(`${this.id}: getTemperature`);
        tempStream.on("data", c => console.log(`${c.device}: Received temp: ${c.celcius} after ${dataStopwatch.lap(false) / 1000} s`));
        tempStream.on("end", () => dataStopwatch.stop());
        tempStream.on("error", err => {
            dataStopwatch.stop(" (failed)");
            console.error(`${this.id}: error:`, err);
        });

        for (let i = 1; i < attempts; i++) {
            await stopwatch.runAsync(`${this.id} attempt ${i+1}`, async () => {
                await BBPromise.delay(secondsBetweenAttemps * 1000);
                celcius = this.changeTemperature(celcius);
                await this.setTemperature(celcius);
            });
        }
        console.log(`${this.id}: ALL DONE`);
    }

    private setTemperature(celcius: number) {
        console.log(`${this.id}: Setting temperature to ${celcius}`);
        return stopwatch.runAsync(
            `${this.id}: setTemperature`,
            () => this.client.temperatureService.setTemperatureAsync({device: this.id, celcius})
        );
    }

    private randomTemperature() {
        return (Math.random() * (Device.MaxTemp - Device.MinTemp)) + Device.MinTemp;
    }

    private changeTemperature(celcius: number) {
        const diff = Math.random();
        const sign = Math.round(Math.random());

        return sign > 0 ? Math.min(celcius + diff, Device.MaxTemp) : Math.max(celcius - diff, Device.MinTemp);
    }
}

function main() {
    const nofDevices = 9;
    const waitSecondsBetweenUpdates = 1;
    const updatesPerDevice = 10;

    const devices: Device[] = new Array(nofDevices);

    for (let i = 0; i < nofDevices; i++) {
        devices[i] = new Device("localhost:50051");
    }
    BBPromise.all(devices.map(d => d.run(updatesPerDevice, waitSecondsBetweenUpdates)));
}

main();

File: src/dotnetcore/LoadTest.cs
using System;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using Device;
using Grpc.Core;

class SimulatedDevice
{
    private static int nextId = 0;
    private const float MinTemp = -10;
    private const float MaxTemp = 45;

    private static readonly Random rnd = new Random();

    private readonly Channel channel;

    private readonly TemperatureService.TemperatureServiceClient client;

    public string ID
    {
        get;
    }

    public SimulatedDevice(string endpoint)
    {
        ID = $"device-{nextId++}";

        channel = new Channel(endpoint, ChannelCredentials.Insecure);
        client = new TemperatureService.TemperatureServiceClient(channel);
    }

    public async Task Run(uint attempts, TimeSpan timeBetweenAttempts)
    {
        var celcius = RandomTemperature();
        var cancelSource = new CancellationTokenSource();

        await SetTemperatureAsync(celcius);
        var getTemps = GetTemperatures(cancelSource.Token);

        for (var i = 1; i < attempts; i++)
        {
            await StopwatchExtension.RunAsync(
                $"{ID}: attempt {i+1}",
                async () => {
                    await Task.Delay(timeBetweenAttempts);
                    celcius = ChangeTemperature(celcius);
                    return await SetTemperatureAsync(celcius);
                }
            );
        }
        Console.WriteLine($"{ID}: done setting temperatures");
        cancelSource.Cancel();
        await getTemps;
        await channel.ShutdownAsync();
        Console.WriteLine($"{ID}: ALL DONE");
    }

    private async Task<Empty> SetTemperatureAsync(float celcius)
    {
        Console.WriteLine($"{ID}: Setting temperature to {celcius}");
        return await StopwatchExtension.RunAsync<Empty>($"{ID} setTemperature {celcius}",
            () => client.set_temperatureAsync(new Temperature {
                Device = ID,
                Celcius = celcius
            }).ResponseAsync
        );
    }

    private async Task GetTemperatures(CancellationToken token)
    {
        var sw = Stopwatch.StartNew();
        using (var call = client.get_temperature(new Device.Device { Device_ = ID }))
        using (var stream = call.ResponseStream)
        {
            // For some reason we cannot use a cancellable token here, so handle cancellation manually...
            while (await stream.MoveNext(CancellationToken.None))
            {
                Console.WriteLine($"{ID}: received temperature {stream.Current.Celcius} (gap: {sw.Elapsed}).");
                sw.Restart();
                if (token.IsCancellationRequested)
                {
                    Console.WriteLine("Cancelling GetTemperature");
                    break;
                }
            }
        }
        Console.WriteLine($"{ID} end of temeperature stream ({sw.Elapsed} since last receive)");
    }

    private float RandomTemperature()
    {
        return ((float) rnd.NextDouble() * (MaxTemp - MinTemp)) + MinTemp;
    }

    private float ChangeTemperature(float celcius) {
        var diff = (float) rnd.NextDouble();
        var sign = rnd.NextDouble() >= .5 ? 1 : -1;

        return Math.Max(MinTemp, Math.Min(MaxTemp, sign * diff));
    }
}

static class StopwatchExtension
{
    public static async Task<T> RunAsync<T>(string name, Func<Task<T>> asyncAction)
    {
        var stopwatch = Stopwatch.StartNew();
        try
        {
            return await asyncAction();
        }
        finally
        {
            stopwatch.Stop();
            Console.WriteLine($"{name} finished, took {stopwatch.Elapsed}.");
        }
    }
}